File: cpp/storage.cpp
#include "storage.h"

#include <QDir>
#include <QFile>
#include <QJsonArray>
#include <QJsonDocument>
#include <QStandardPaths>
#include <QVariantList>

StorageManager::StorageManager(const QString &baseDir)
{
    if (!baseDir.isEmpty()) {
        m_baseDir = baseDir;
    } else {
        const QByteArray xdg = qgetenv("XDG_DATA_HOME");
        if (!xdg.isEmpty())
            m_baseDir = QString::fromLocal8Bit(xdg) + QStringLiteral("/omarchy-sudoku");
        else
            m_baseDir = QDir::homePath() + QStringLiteral("/.local/share/omarchy-sudoku");
    }
    QDir().mkpath(m_baseDir);
    m_saveFile = m_baseDir + QStringLiteral("/savegame.json");
    m_highscoreFile = m_baseDir + QStringLiteral("/highscores.json");
}

bool StorageManager::hasSavedGame() const
{
    if (!QFile::exists(m_saveFile))
        return false;
    QFile f(m_saveFile);
    if (!f.open(QIODevice::ReadOnly))
        return false;
    const QJsonDocument doc = QJsonDocument::fromJson(f.readAll());
    if (!doc.isObject())
        return false;
    const QJsonObject data = doc.object();
    const QJsonArray board = data.value(QStringLiteral("board")).toArray();
    const QJsonArray solution = data.value(QStringLiteral("solution")).toArray();
    if (board.size() != 81 || solution.size() != 81)
        return false;

    bool finished = true;
    for (int i = 0; i < 81; ++i) {
        const int b = board[i].toInt();
        const int s = solution[i].toInt();
        if (!(b == s && b != 0)) {
            finished = false;
            break;
        }
    }
    if (finished)
        return false;

    const QJsonArray initialClues = data.value(QStringLiteral("initial_clues")).toArray();
    const QJsonArray notes = data.value(QStringLiteral("notes")).toArray();
    const int timeSpent = data.value(QStringLiteral("time")).toInt(0);
    if (initialClues.size() == 81) {
        bool hasMoves = false;
        for (int i = 0; i < 81; ++i) {
            if (board[i].toInt() != 0 && !initialClues[i].toBool()) {
                hasMoves = true;
                break;
            }
        }
        bool hasNotes = false;
        for (const QJsonValue &n : notes) {
            if (n.toArray().size() > 0) {
                hasNotes = true;
                break;
            }
        }
        if (!(hasMoves || hasNotes || timeSpent > 10))
            return false;
    }
    return true;
}

void StorageManager::saveGame(const QJsonObject &state)
{
    QFile temp(m_saveFile + QStringLiteral(".tmp"));
    if (!temp.open(QIODevice::WriteOnly))
        return;
    temp.write(QJsonDocument(state).toJson(QJsonDocument::Indented));
    temp.close();
    QFile::remove(m_saveFile);
    QFile::rename(temp.fileName(), m_saveFile);
}

std::optional<QJsonObject> StorageManager::loadGame() const
{
    if (!QFile::exists(m_saveFile))
        return std::nullopt;
    QFile f(m_saveFile);
    if (!f.open(QIODevice::ReadOnly))
        return std::nullopt;
    const QJsonDocument doc = QJsonDocument::fromJson(f.readAll());
    if (!doc.isObject())
        return std::nullopt;
    return doc.object();
}

void StorageManager::deleteSavedGame()
{
    if (QFile::exists(m_saveFile))
        QFile::remove(m_saveFile);
}

QVariantMap StorageManager::getAllHighscores() const
{
    QVariantMap scores{
        {QStringLiteral("simple"), 0},
        {QStringLiteral("easy"), 0},
        {QStringLiteral("intermediate"), 0},
        {QStringLiteral("expert"), 0},
    };
    if (!QFile::exists(m_highscoreFile))
        return scores;
    QFile f(m_highscoreFile);
    if (!f.open(QIODevice::ReadOnly))
        return scores;
    const QJsonDocument doc = QJsonDocument::fromJson(f.readAll());
    if (!doc.isObject())
        return scores;
    const QJsonObject data = doc.object();
    for (auto it = scores.begin(); it != scores.end(); ++it)
        it.value() = data.value(it.key()).toInt(0);
    return scores;
}

int StorageManager::getHighscore(const QString &difficultyKey) const
{
    return getAllHighscores().value(difficultyKey, 0).toInt();
}

bool StorageManager::setHighscore(const QString &difficultyKey, int score)
{
    QVariantMap scores = getAllHighscores();
    const int current = scores.value(difficultyKey, 0).toInt();
    if (score <= current)
        return false;
    scores[difficultyKey] = score;
    QJsonObject obj;
    for (auto it = scores.constBegin(); it != scores.constEnd(); ++it)
        obj.insert(it.key(), it.value().toInt());
    QFile temp(m_highscoreFile + QStringLiteral(".tmp"));
    if (!temp.open(QIODevice::WriteOnly))
        return false;
    temp.write(QJsonDocument(obj).toJson(QJsonDocument::Indented));
    temp.close();
    QFile::remove(m_highscoreFile);
    QFile::rename(temp.fileName(), m_highscoreFile);
    return true;
}

File: src/storage.py
"""
Persistence manager for Sudoku game state and highscores.
Stores JSON files in ~/.local/share/omarchy-sudoku/.
"""

import json
import os
from pathlib import Path
from typing import Any, Optional


class StorageManager:
    def __init__(self, base_dir: Optional[Path | str] = None):
        if base_dir:
            self.base_dir = Path(base_dir)
        else:
            data_dir = os.environ.get("XDG_DATA_HOME")
            if data_dir:
                self.base_dir = Path(data_dir) / "omarchy-sudoku"
            else:
                self.base_dir = Path.home() / ".local" / "share" / "omarchy-sudoku"

        self.base_dir.mkdir(parents=True, exist_ok=True)
        self.save_file = self.base_dir / "savegame.json"
        self.highscore_file = self.base_dir / "highscores.json"

    def has_saved_game(self) -> bool:
        if not self.save_file.is_file():
            return False
        try:
            with open(self.save_file, "r", encoding="utf-8") as f:
                data = json.load(f)
            if not isinstance(data, dict):
                return False
            board = data.get("board", [])
            solution = data.get("solution", [])
            if len(board) != 81 or len(solution) != 81:
                return False
            # If all cells are filled and match solution, game is already finished
            if all(board[i] == solution[i] and board[i] != 0 for i in range(81)):
                return False
            # Check if game was actually played (at least one user entry, note, or >10s played)
            initial_clues = data.get("initial_clues", [])
            notes = data.get("notes", [])
            time_spent = data.get("time", 0)
            if initial_clues and len(initial_clues) == 81:
                has_moves = any(board[i] != 0 and not initial_clues[i] for i in range(81))
                has_notes = any(len(n) > 0 for n in notes) if notes else False
                if not (has_moves or has_notes or time_spent > 10):
                    return False
            return True
        except Exception:
            return False

    def save_game(self, state: dict[str, Any]) -> None:
        try:
            temp_file = self.save_file.with_suffix(".tmp")
            with open(temp_file, "w", encoding="utf-8") as f:
                json.dump(state, f, indent=2)
            temp_file.replace(self.save_file)
        except Exception as e:
            print(f"Error saving game: {e}")

    def load_game(self) -> Optional[dict[str, Any]]:
        if not self.save_file.is_file():
            return None
        try:
            with open(self.save_file, "r", encoding="utf-8") as f:
                return json.load(f)
        except Exception as e:
            print(f"Error loading saved game: {e}")
            return None

    def delete_saved_game(self) -> None:
        if self.save_file.exists():
            try:
                self.save_file.unlink()
            except Exception as e:
                print(f"Error deleting saved game: {e}")

    def get_all_highscores(self) -> dict[str, int]:
        if not self.highscore_file.is_file():
            return {"simple": 0, "easy": 0, "intermediate": 0, "expert": 0}
        try:
            with open(self.highscore_file, "r", encoding="utf-8") as f:
                data = json.load(f)
                return {
                    "simple": int(data.get("simple", 0)),
                    "easy": int(data.get("easy", 0)),
                    "intermediate": int(data.get("intermediate", 0)),
                    "expert": int(data.get("expert", 0)),
                }
        except Exception as e:
            print(f"Error reading highscores: {e}")
            return {"simple": 0, "easy": 0, "intermediate": 0, "expert": 0}

    def get_highscore(self, difficulty_key: str) -> int:
        scores = self.get_all_highscores()
        return scores.get(difficulty_key, 0)

    def set_highscore(self, difficulty_key: str, score: int) -> bool:
        """Update highscore if greater. Returns True if a new highscore was set."""
        scores = self.get_all_highscores()
        current = scores.get(difficulty_key, 0)
        if score > current:
            scores[difficulty_key] = score
            try:
                temp_file = self.highscore_file.with_suffix(".tmp")
                with open(temp_file, "w", encoding="utf-8") as f:
                    json.dump(scores, f, indent=2)
                temp_file.replace(self.highscore_file)
                return True
            except Exception as e:
                print(f"Error writing highscores: {e}")
        return False
